DeepSpeech.ipynb

```diff
@@ -1337,6 +1362,44 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
+    "Finally, export the model."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "with tf.device('/cpu:0'):\n",
+    "    model_exporter = exporter.Exporter(saver)\n",
+    "    \n",
+    "    #print [n.name for n in session.graph.as_graph_def().node]\n",
+    "    \n",
+    "    inputs = {}\n",
+    "    outputs = {}\n",
+    "    for i in xrange(len(available_devices)):\n",
+    "        # Create output for tower i\n",
+    "        input = session.graph.get_tensor_by_name('tower_%d/input_batch:0' % i)\n",
+    "        output = session.graph.get_tensor_by_name('tower_%d/output_decoded:0' % i)\n",
+    "        inputs['tower_%d/input_batch' % i] = input\n",
+    "        outputs['tower_%d/result' % i] = output\n",
+    "\n",
+    "    model_exporter.init(session.graph.as_graph_def(),\n",
+    "                        named_graph_signatures = {\n",
+    "                            'inputs': exporter.generic_signature(inputs),\n",
+    "                            'outputs': exporter.generic_signature(outputs)})\n",
```

Comment:
More summary/recap from IRC:

We want to have a single input tensor, but currently batch fetching happens asynchronously with training. If we have a single input tensor, it means all the data must be available for that tensor (or we need to use some kind of feeding mechanism that I'm not aware of - which more than likely exists, but no one has mentioned it and I've not seen it yet). Otherwise before a training iteration can begin, all devices will need to wait for the all batches to be fetched before beginning work. Currently, training on a batch can happen immediately after it's available.

This PR works around that by having separate inputs and outputs per device (probably - though the export seems to work, the resulting model is currently untested as I've not yet written the client). We may want to investigate other feeding mechanisms, but I believe that the changes required to have one input tensor without causing any extra blocking would be fairly significant. I'm not sure, given my limited understanding of both TensorFlow and this notebook, that I'd be the best person to do that in a short timeframe.